src/main/java/frc/robot/subsystems/elevator/ElevatorIOReal.java

```diff
@@ -1,15 +1,96 @@
+    private final TalonFX mainMotor;
+    private final TalonFX auxMotor;
+
+    private final DigitalInput bottomSensor;
+    private final DigitalInput topSensor;
+
+    private final MotionMagicVoltage positionControl = new MotionMagicVoltage(0);
+    private final DutyCycleOut powerControl = new DutyCycleOut(0);
+
+    public ElevatorIOReal() {
+        mainMotor = new TalonFX(Ports.Elevator.MAIN_MOTOR);
+        auxMotor = new TalonFX(Ports.Elevator.AUX_MOTOR);
+        servo = new Servo(Ports.Elevator.SERVO);
+        bottomSensor = new DigitalInput(Ports.Elevator.BOTTOM_LIMIT_SWITCH);
+        topSensor = new DigitalInput(Ports.Elevator.TOP_LIMIT_SWITCH);
+
+        mainMotor.getConfigurator().apply(ElevatorConstants.MOTOR_CONFIGURATION);
+        auxMotor.getConfigurator().apply(ElevatorConstants.MOTOR_CONFIGURATION);
+
+        auxMotor.setControl(new StrictFollower(mainMotor.getDeviceID()));
+    }
+
     @Override
-    public void setPower(double power) {}
+    public void setPower(double power) {
+        mainMotor.setControl(powerControl.withOutput(power));
+    }
 
     @Override
-    public void setHeight(MutableMeasure<Distance> height) {}
+    public void setHeight(MutableMeasure<Distance> height) {
+        inputs.heightSetpoint.mut_replace(height);
+        mainMotor.setControl(positionControl.withPosition(height.in(Units.Meters)));
+    }
 
     @Override
-    public void updateInputs(ElevatorInputs inputs) {}
+    public void updateInputs(ElevatorInputs inputs) {
+        inputs.isBottom = atBottom();
+        inputs.isTop = atTop();
+        inputs.carriageHeight = Meters.of(mainMotor.getPosition().getValue()).mutableCopy();
+
+        inputs.gripperHeight.mut_replace(
+                inputs.carriageHeight.gt(ElevatorConstants.GRIPPER_HEIGHT)
+                        ? inputs.carriageHeight
+                                .mutableCopy()
+                                .mut_minus(ElevatorConstants.GRIPPER_HEIGHT)
+                        : Meters.zero());
+
+        inputs.servoAngle = Units.Degrees.of(servo.getAngle()).mutableCopy();
+    }
+
+    public boolean atTop() {
+        return topSensor.get();
+    }
+
+    public boolean atBottom() {
+        return bottomSensor.get();
+    }
+
+    public void flipServo() {
+        if (servo.get() == ElevatorConstants.SERVO_OPEN.in(Degrees)) {
+            inputs.servoSetpoint = ElevatorConstants.SERVO_CLOSE;
+            servo.set(ElevatorConstants.SERVO_CLOSE.in(Degrees));
+        } else {
+            inputs.servoSetpoint = ElevatorConstants.SERVO_OPEN;
+            servo.set(ElevatorConstants.SERVO_OPEN.in(Degrees));
+        }
+    }
+
+    public void resetEncoder() {
+        if (atTop()) {
+            setHeight(Meters.of(0).mutableCopy());
+        } else if (atTop())
+            setHeight(Meters.of(ElevatorConstants.MAX_HEIGHT.in(Meters)).mutableCopy());
+    }
+
+    public void stopMotor() {
+        mainMotor.setControl(powerControl.withOutput(0));
```

Comment:
There's a shorter method that does that. Look it up in the API reference.

Reply:
d7f3e349b3c60f3c884c865911f2ffa3fb78b21f

Reply:
Nice!